overhaul vat-timer: virtualized, durable, upgradable, Clock, TimerBrand

warner opened the pull request:
vat-timer is now fully virtualized, durablized, and upgradeable. RAM
usage should be O(N) in the number of:

* pending Promise wakeups (`wakeAt`, `delay`)
* active Notifier promises (`makeNotifier`)
* active Iterator promises (`makeNotifier()[Symbol.asyncIterator]`)

Pending promises will be disconnected (rejected) during upgrade, as
usual.

All handlers and Promises will fire with the most recent timestamp
available, which (under load) may be somewhat later than the scheduled
wakeup time.

Until cancellation, Notifiers will always report a scheduled time
(i.e. `start` plus some multiple of the interval). The opaque
`updateCount` used in Notifier updates is a counter starting from 1n.
When a Notifier is cancelled, the final/"finish" value is the
timestamp of cancellation, which may or may not be a multiple of the
interval (and might be a duplicate of the last non-final value). Once
in the cancelled state, `getUpdateSince(anything)` yields `{ value:
cancellationTimestamp, updateCount: undefined }`, and the
corresponding `iterator.next()` resolves to `{ value:
cancellationTimestamp, done: true }`. Neither will ever reject their
Promises (except due to upgrade).

Asking for a wakeup in the past or present will fire immediately.

Most API calls will accept an arbitrary Far object as a CancelToken,
which can be used to cancel the wakeup/repeater. `makeRepeater` is the
exception.

This does not change the device-timer API or implementation, however
vat-timer now only uses a single device-side wakeup, and only exposes
a single handler object, to minimize the memory usage and object
retention by the device (since devices do not participate in GC).

This introduces a `Clock` which can return time values without also
providing scheduling authority, and a `TimerBrand` which can validate
time values without providing clock or scheduling
authority. Timestamps are not yet Branded, but the scaffolding is in
place.

`packages/SwingSet/tools/manual-timer.js` offers a manually-driven
timer service, which can help with unit tests.

closes #4282
refs #4286
closes #4296
closes #5616
closes #5668
closes #5709
refs #5798

erights reviewed (commented):
Just some comments so far. #5862 is also a review comment.

I have not yet dug into the meat of this. Looking forward to doing so.

warner commented on packages/SwingSet/package.json:
```diff
@@ -31,6 +31,7 @@
     "@agoric/notifier": "^0.4.0",
     "@agoric/store": "^0.7.2",
     "@agoric/swing-store": "^0.7.0",
+    "@agoric/vat-data": "^0.3.1",
```
Oh, yeah, Chip and Richard and I talked about that.

vat-data has a `devDependencies` on swingset, to support its unit tests, but not a `dependencies`. vat-data looks for names on `globalThis`, which are usually there because swingset added it to the vat environment, but vat-data doesn't actually import anything from swingset.

So I think we don't have a real package-level cycle: we have a package-level `devDependency` cycle, which I think is ok.

erights replied:
I'd forgotten about the global thing. Weird way to depend de facto but not de jure ;). devDependencies are indeed ok, so all is good here. Resolving.

warner commented on packages/SwingSet/src/vats/timer/types.d.ts:
```diff
@@ -88,12 +88,6 @@ declare global {
      * Return value is the time at which the call is scheduled to take place
      */
```
yeah, I haven't tried to improve `types.d.ts` at all, I just tried to make sure the typechecker wasn't complaining. I'll make a pass through this file and try to remove the stale parts.

warner commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,854 @@
+ * @typedef WakeupPromiseControls
+ * @property {(scheduled: Timestamp) => void} resolve
+ * @property {(err: unknown) => void} reject
+ * @typedef {LegacyWeakMap<PromiseEvent, WakeupPromiseControls>} WakeupPromiseTable
```
Oh, it looks like I could store the resolve/reject functions in a Far (even though they're never revealed off-vat), and that would let us use a `WeakMapStore` instead of the `LegacyWeakMap`. I'll do that.

The reason we can'ut use a (virtual) `BigWeakMapStore` is that we don't have virtual/durable Promises. The #4567 "durable notifier" plan will need this pattern too, to hold both the Notifier's `Promise` and it's resolve/reject control functions.

erights replied:
I see it's done. Resolving.

warner commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,854 @@
+ * callbacks while "executing", and by mostly nothing when "cancelled".
+ */
+
+// these internal functions are exported for unit tests
```
I'll add the "this is why this is safe" comment. They're pure functions. Also, they're exported under a separate name, so when this file is used as a vat entry point, we can leverage the fact that liveslots and the `subprocess-xsnap` supervisor don't touch anything other than `buildRootObject`.

For the benefit of the `manualTimer` unit-test support tool (used by zoe's tests, and others), I might need to add a non-pure portion which lets the unit test driver get a callback with every `wake()` result promise, so it can `await` them as it sees fit. I might also be able to get away without that.. still trying to look for solutions. If I do add that, we'll be relying upon the "only `buildRootObject`" aspect for security.

warner commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,854 @@
+function removeEvent(schedule, when, event) {
+  if (schedule.has(when)) {
+    /** @typedef { Event[] } */
+    const oldEvents = schedule.get(when);
+    /** @typedef { Event[] } */
+    const newEvents = oldEvents.filter(ev => ev !== event);
```
hm, it's "before removing" and "after removing".. maybe I'll try a `originalEvents` and `remainingEvents`, if it doesn't make the line too long

erights commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,854 @@
+  // it ourselves.
+
+  /** @type {WakeupPromiseTable} */
+  const wakeupPromises = makeLegacyWeakMap('promises');
```
I see it's done. Resolving.

erights commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,854 @@
+  function wakeAtInternal(when, cancelToken) {
+    const event = makePromiseEvent(when, cancelToken);
+    const { resolve, reject, promise } = makePromiseKit();
+    const controls = harden({ resolve, reject });
```
I see it's done. Resolving.

erights commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,854 @@
+
+  function noContext(f) {
+    return (context, ...args) => f(...args);
+  }
```
Gone. Resolving.

erights commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,830 @@
+  // keys, and upon our Stores performing efficient iteration.
+
+  /** @type {Schedule} */
+  const schedule = provide(baggage, 'schedule', () =>
+    makeScalarBigMapStore('schedule', { durable: true }),
+  );
```
```suggestion
  const schedule = provideDurableMapStore(baggage, 'schedule');
```

erights commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,830 @@
+  // and remove it from 'schedule'
+
+  /** @type {CancelTable} */
+  const cancels = provide(baggage, 'cancels', () =>
+    makeScalarBigWeakMapStore('cancels', { durable: true }),
+  );
```
```suggestion
  const cancels = provideDurableWeakMapStore(baggage, 'cancels');
```

erights commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,830 @@
+  // define behaviors for these. We are also obligated to support the
+  // vivifySingleton calls at the bottom: wakeupHandler, timerService,
+  // timerClock, and timerBrand.
+  const oneShotEventHandle = provideKindHandle(baggage, 'oneShotEventHandle');
+  const promiseEventHandle = provideKindHandle(baggage, 'promiseEventHandle');
+  const repeaterEventHandle = provideKindHandle(baggage, 'repeaterEventHandle');
+  const repeaterHandle = provideKindHandle(baggage, 'repeaterHandle');
+  const notifierHandle = provideKindHandle(baggage, 'notifierHandle');
```
For every case where we have a call to `provideKindHandle` and a corresponding call to `defineDurableKind` or `defineDurableKindMulti`, I think the code would be clearer to replace these with a single call to `vivifyKind` or `vivifyKindMulti`. Typically, programs will have no other uses of kind handles, e.g., no direct calls to `makeKindHandle`. By making this change, those programs hide the entire concept of a kindHandle, making the durability API easier to learn, and the code using it easier to understand.

For me this was one of the punchlines of seeing our own experimental use of the durability/upgrade API change over time. We gradually climbed abstractions until we got to `provide*` and `vivify*`, therefore initially thinking of them as helpers for saying the other things more compactly. But we now hope most upgradable code can rely exclusively on `provide*` and `vivify*`, typically hiding the need to think about the concepts they encapsulate.

erights reviewed (commented):
Just more comments so far. Not done reviewing yet.

erights commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,830 @@
+  }
+
+  function reschedule() {
+    assert(wakeupHandler, 'reschedule() without wakeupHandler');
```
```suggestion
```

This `assert` is no longer meaningful, now that `wakeupHandler` is a const.

erights reviewed (approved):
LGTM!

There are still things I'd like to understand better. But we can discuss them after this is merged (and once I get back on Wednesday).

warner commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,854 @@
+ * @typedef Handler
+ * @property {(scheduled: Timestamp) => unknown} wake
+ * @typedef {unknown} CancelToken
+ * // CancelToken must be pass-by-reference and durable, either local or remote
```
I wasn't sure what the type parser was sensitive to.. I'll remove the `//`

warner commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,854 @@
+
+  // -- helper functions
+
+  /**
+   * convert an internal TimestampValue into a branded Timestamp
+   *
+   * @param {TimestampValue} when
+   * @returns {Timestamp}
+   */
+  function toTimestamp(when) {
+    return TimeMath.toAbs(when); // TODO (when, brand)
+  }
+
+  /**
+   * convert external (branded) Timestamp to internal bigint
+   *
+   * @param {Timestamp} when
+   * @returns {TimestampValue}
+   */
+  function fromTimestamp(when) {
+    return TimeMath.absValue(when); // TODO: brand
+  }
+
+  /**
+   * convert external (branded) RelativeTime to internal bigint
+   *
+   * @param {RelativeTime} delta
+   * @returns {RelativeTimeValue}
+   */
+  function fromRelativeTime(delta) {
+    return TimeMath.relValue(delta);
+  }
```
I will pretend that I was thinking that far ahead :)

warner commented:
Questions for reviewers:
* Somebody said that the new `timerBrand` did not need a `isMyClock()` method, but now I'm not so sure. If you hold a branded Timestamp and someone hands you a Clock, would you ever want to verify that the Clock produces the right kind of timestamps? It seems to me that if you aren't going to be handed a TimerService too, there's no way to detect a deceptive Clock without this extra method.
* when Promise-flavored timer/repeater/notifier is cancelled, this rejects any Promise with `Error('TimerCancelled')`
  * should I use a non-Error rejection object instead, like `{ name: 'TimerCancelled' }` ? We do that with Promises that are rejected because of a vat upgrade, so clients can distinguish those cancellations from actual rejections, but I'm not sure that's appropriate here
  * that Error causes a `vat-timer.js` stack trace to be included in the rejection (userspace does not see it, because of our censorship, but it's noisy in unit tests, and implies something went wrong within vat-timer, when cancellation is entirely controlled by clients). Would it be appropriate for me to create that `Error` object once, at top-level module scope, and re-use the error singleton whenever a rejection is needed? OTOH I suppose that would capture the stack of liveslots and the supervisor when they first import the timer vat code, which is even less relevant.
* All APIs (except for Notifiers and their iterators) now fire with the most recent timestamp, not the time they were originally scheduled for (these would only diverge if the timer device and timer vat suffered delays in communication).
  * Notifiers (and their iterators) fire with the most recent nominal event prior to the query. They simulate an "eager" Notifier that is being triggered every `interval` ticks, from which the client is fetching the most recent update. If start a Notifier with `delay=0` and manage to submit your `getUpdateSince(undefined)` before any time has passed, you'll get a Promise that fulfills promptly to the current time.
  * I *think* this is mostly the same as the previous behavior, but the Zoe manual timer behaved slightly differently, so I wanted to make sure this specified behavior seems correct
* I replaced Zoe's `manualTimer.js` with a logging/awaiting-augmented wrapper around a new SwingSet-provided `manual-timer.js` (which only has `advanceTo(newTime)`)
  * some of the "&& scheduling event for.." log messages were too hard to support, so I removed them from the test cases that expected them
  * the delivery order of some nominally-simultaneous events seems to have changed, so a few "golden transcript" -style tests were modified to match
  * the `fakePriceAuthority` needed to start its Repeater one tick later, to get the same results as before. I'm not entirely sure how this works, so I'm hoping someone else can look at it and make sure I didn't break anything

There is a general problem of how to know when "background" async work has finished, before it is safe to either check for consequences, or to initiate further async activity that effectively polls mutable state that the background work might or might not have finished modifying. The old Zoe `manualTimer.js` offered some synchronization tools, but the new (real) timer API is bigger and doesn't lend itself to as complete a solution. I've added enough promise-tracking `await`s into the Zoe wrapper to support the existing tests (with some small changes), but in general, new tests should probably use `await eventLoopIteration()` (or some other `setImmediate` -based wrapper, *not* just a few `await Promise.resolve()`s) after calling `tick()`, if they want to ensure that all the wakeups thus triggered have finally rolled to a halt.

This is not something the `manualTimer` can do internally, because several zoe/etc `test/swingsetTests/` use the manual timer *inside a vat*, where `setImmediate` is not available. I changed `tickN()` to accept `eventLoopIteration` as an argument, which it will call and `await` if present, so the "unit" unit-tests can do that, while the "swingset" unit-tests do not.

I have not yet changed the TimerService API to use Brand-bearing timestamps, but the cutover will be just a few lines in `vat-timer.js`, and I'm guessing we should do that very soon. The swingset unit tests will require more work to adapt, but it's mostly mechanical.

Chris-Hibbert reviewed (commented):
some minor proofreading comments.

I'm done with vat-timer and its primary tests, but have a bunch of other code to review still.

erights commented:
> Questions for reviewers:
> 
> * Somebody said that the new `timerBrand` did not need a `isMyClock()` method, but now I'm not so sure. If you hold a branded Timestamp and someone hands you a Clock, would you ever want to verify that the Clock produces the right kind of timestamps? It seems to me that if you aren't going to be handed a TimerService too, there's no way to detect a deceptive Clock without this extra method.

That reasoning sounds right to me.

> * when Promise-flavored timer/repeater/notifier is cancelled, this rejects any Promise with `Error('TimerCancelled')`
>   
>   * should I use a non-Error rejection object instead, like `{ name: 'TimerCancelled' }` ? We do that with Promises that are rejected because of a vat upgrade, so clients can distinguish those cancellations from actual rejections, but I'm not sure that's appropriate here
>   * that Error causes a `vat-timer.js` stack trace to be included in the rejection (userspace does not see it, because of our censorship, but it's noisy in unit tests, and implies something went wrong within vat-timer, when cancellation is entirely controlled by clients). Would it be appropriate for me to create that `Error` object once, at top-level module scope, and re-use the error singleton whenever a rejection is needed? OTOH I suppose that would capture the stack of liveslots and the supervisor when they first import the timer vat code, which is even less relevant.

When the throw or rejection is not indicating that something bad happened, not indicating that a possible bug symptom may have just occurred, then I prefer that the thrown thing or the rejection reason not be an error, for all the reasons you state above. I like your suggestion that it should be strongly analogous, while testably distinct, from the rejection reasons with which we indicate upgrade. `harden({ name: 'TimerCancelled' })` seems like a fine choice to me.

> * All APIs (except for Notifiers and their iterators) now fire with the most recent timestamp, not the time they were originally scheduled for (these would only diverge if the timer device and timer vat suffered delays in communication).

This seems right.

>   * Notifiers (and their iterators) fire with the most recent nominal event prior to the query. They simulate an "eager" Notifier that is being triggered every `interval` ticks, from which the client is fetching the most recent update. If start a Notifier with `delay=0` and manage to submit your `getUpdateSince(undefined)` before any time has passed, you'll get a Promise that fulfills promptly to the current time.
>   * I _think_ this is mostly the same as the previous behavior, but the Zoe manual timer behaved slightly differently, so I wanted to make sure this specified behavior seems correct

In the scenario you mention, "...`delay=0`...before any time has passed...", would the old notifier wait a tick first? Neither behavior seems incorrect to me, but the new behavior seems better.

> * I replaced Zoe's `manualTimer.js` with a logging/awaiting-augmented wrapper around a new SwingSet-provided `manual-timer.js` (which only has `advanceTo(newTime)`)
>   * some of the "&& scheduling event for.." log messages were too hard to support, so I removed them from the test cases that expected them
>   * the delivery order of some nominally-simultaneous events seems to have changed, so a few "golden transcript" -style tests were modified to match

Sure. Goldens are fragile, and the notifier spec is inherently non-deterministic.

>   * the `fakePriceAuthority` needed to start its Repeater one tick later, to get the same results as before. I'm not entirely sure how this works, so I'm hoping someone else can look at it and make sure I didn't break anything
> 
> There is a general problem of how to know when "background" async work has finished, before it is safe to either check for consequences, or to initiate further async activity that effectively polls mutable state that the background work might or might not have finished modifying. The old Zoe `manualTimer.js` offered some synchronization tools, but the new (real) timer API is bigger and doesn't lend itself to as complete a solution. I've added enough promise-tracking `await`s into the Zoe wrapper to support the existing tests (with some small changes), but in general, new tests should probably use `await eventLoopIteration()` (or some other `setImmediate` -based wrapper, _not_ just a few `await Promise.resolve()`s) after calling `tick()`, if they want to ensure that all the wakeups thus triggered have finally rolled to a halt.
> 
> This is not something the `manualTimer` can do internally, because several zoe/etc `test/swingsetTests/` use the manual timer _inside a vat_, where `setImmediate` is not available. I changed `tickN()` to accept `eventLoopIteration` as an argument, which it will call and `await` if present, so the "unit" unit-tests can do that, while the "swingset" unit-tests do not.

I'm not following this yet.

> I have not yet changed the TimerService API to use Brand-bearing timestamps, but the cutover will be just a few lines in `vat-timer.js`, and I'm guessing we should do that very soon. The swingset unit tests will require more work to adapt, but it's mostly mechanical.

Good. There's no need to do that in this same PR.

warner commented on packages/SwingSet/docs/timer.md:
```diff
@@ -67,42 +67,86 @@ A single application might have multiple sources of time, which would require th
+  // every 60 seconds thereafter. The next wakeup will not be scheduled
+  // until the handler message is acknowledged (when its return promise is
+  // fulfilled), so wakeups might be skipped, but they will always be
+  // scheduled for the next 'start + k * repeat', so they will not
```
aye, but maybe `now + delay + k * interval`

warner commented:
It looks like 1: zoe's `buildManualTimer` is being used by dapps too, so any API changes must do argument-sensing and be backwards compatible, and 2: the "`fakePriceAuthority` has to start at 1 instead of 0" problem that I saw in zoe will also occur in the dapp tests that have copied that code. So I think I have to dig into the dapps and understand that fencepost problem before I can land this without disruption, sigh.

warner commented:
Today's discussion pointed out some changes that I need to make:

* change TimerNotifier to use an `updateCount` that starts at `1n` and increments sequentially from there, by using something like `BigInt(Math.floor((now - delay) / interval)+1)`
  * @erights was weirded out by `updateCount` not being sequential
  * but this would also simplify the conditionals in `getUpdateSince`
  * also remove the `Error('invalid updateCount for timer Notifier')` clause
  * if the supplied `updateCount` is not `undefined`, then we check whether it matches the current value, or is different: don't do more than that
  * the `+1` is to ensure that a live `updateCount` is not falsy, to guard against clients who accidentally use `if (!updateCount)` instead of `if (updateCount === undefined)` to sense the finish condition
* change TimerNotifier's canceller to put the notifier into a "finish" state, rather than an erroring state
  * cancellation should record the current time, to compute a final update value
  * this makes all outstanding and all future `getUpdateSince` promises fire with `{ value: lastTime, updateCount: lastUpdateCount }`
* zoe's `buildManualTimer` will change from the old `buildManualTimer(log, startTime=0n, timeStep=1n)` to `buildManualTimer(log=dummy, startTime=0n, options={ timeStep=1n, eventLoopIteration }`
  * if `eventLoopIteration` is provided, `tick()` will finish with `await eventLoopIteration()`
  * this removes the `{ doWait: true }` I was using
  * unit tests will need to change their `buildManualTimer` calls to provide `eventLoopIteration`, but then they won't need to do their own call after tick()
* zoe's `fakePriceAuthority` contains a race because it has multiple uncoordinated calls into the timer, specifically `startTicker` vs `quoteAtTime`
  * the plan is to make `quoteAtTime` follow the `ticker` (the notifier driven by `startTicker`), rather than using `E(timer).setWakeup` itself
  * so `quoteAtTime` looks like:
```js
if (timeStamp <= latestTick) {
  return resolve(priceInQuote(amountIn, brandOut, latestTick));
}
// else push `resolve` into a map that is checked within the `startTicker` handler
```

We walked through the simplest `test-fakePriceAuthority.js` test (`priceAuthority quoteAtTime`), and we think the fourth call to `tick()` was a workaround for the lack of complete `eventLoopIteration` synchronication: the `time = 3n` update (which is odd, not even, so the price should be 55) should be fired by the third call to `tick()`. The failures I've seen in my branches are because the race between the `quoteAtTime` call to `E(timer).setWakeup` and the `startTicker` use of `E(timer).setRepeater`. And `setRepeater(0)` now fires earlier, which doesn't change latestTick but *does* start incrementing `currentPriceIndex` earlier.

I see calls to `buildManualTimer` in three external repositories: `dapp-card-store`, `dapp-oracle`, and `dapp-otc`. None of them use the `timeStep` argument, so it should be relatively safe to convert that argument slot into an options bag.

I see `fakePriceAuthority` mentioned in two external repos: `dapp-oracle` and `dapp-treasury`. The `dapp-oracle` copy is significantly different than the one in zoe: I see things like `makeLinearPriceAuthority` and `makeSinglePriceAuthority`. I don't know if that makes it newer or older than the zoe one, but it will probably require its own analysis. I haven't looked at the `dapp-treasury` copy yet, but it looks like it's in an example, not the main code.

warner commented on packages/SwingSet/test/vat-timer-upgrade/test-vat-timer-upgrade.js:
```diff
@@ -0,0 +1,173 @@
+import { initializeSwingset, makeSwingsetController } from '../../src/index.js';
+import { buildTimer } from '../../src/devices/timer/timer.js';
+
+const bfile = name => new URL(name, import.meta.url).pathname;
```
I think so? I copied this from something I wrote a long time ago, and I was so terse I forget what it was an abbreviation for.

warner commented:
That commit (ce2347b49) implements the Notifier/Iterator changes. For clarity: the "finish state" of a cancelled TimerNotifier has a `value` of the cancellation time, not the latest update (which wouldn't exist anyways, if the Notifier was cancelled before the first event happened).

I'm still working on the `manualTimer` and `priceAuthority` changes, so zoe tests are probably borked.

warner commented:
I think I understand the fencepost error. In the old version (on current trunk):

* Zoe's `manualTimer.makeRepeater(delay=0)` does not wake right away (while `time = 0`). Instead, the first `wake()` occurs after the test calls `tick()` for the first time, and the handlers get a `wake(time=1)`.
* `fakePriceAuthority` has a `handler` that is subscribed to that `makeRepeater(delay=0)`
  * the first time `handler.wake` is called, it leaves `currentPriceIndex` alone (it remains as 0)
  * all subsequent times, it increments `currentPriceIndex`
  * it records the current timestamp in `latestTick`, then updates a Notifier named `ticker` with the timestamp
    * `ticker` is consumed by `generateQuotes`
* some of the API is pretty immediate: `quoteGiven`/`quoteWanted` grabs the current timestamp and uses `currentPriceIndex` to look up the current price, then returns a quote
* `quoteWhenGTE`/etc push a resolver onto `comparisonQueue`, and each time `handler.wake` fires, it resolves anything in the queue that is matched by the new price
* `quoteAtTime` uses its own `E(timer).setWakeup(timeStamp)` to wake a handler at the requested time. It then takes the handler's `time` argument and uses `priceInQuote()` to build a quote from the current `currentPriceIndex` and `time`

The main problem was that `quoteAtTime` is scheduling a wakeup for the same time as the `makeRepeater`, and the invocation order of two such "simultaneous" alarms is not a well-specified part of the TimerService API. The `test-fakePriceAuthority.js` test named `priceAuthority quoteAtTime` was accidentally dependent upon a specific ordering of events:

* the test calls `E(priceAuthority).quoteAtTime(when=3n)`, which schedules a `wakeAt(3n)`
* `tick()` is called several times, all of which fire the `makeRepeater` handler in `startTicker`
  * the `tick()` that advances time from `0n` to `1n` leaves `currentPriceIndex` at 0, and sets `latestTick = 1n`
  * the `tick()` that advances time from `1n` to `2n` updates `currentPriceIndex` to 1, and sets `latestTick = 2n`
* the `wakeAt(3n)` handler fires first, is handed `time=3n`, and `priceInQuote()` is called with `time=3n`
  * it polls `currentPriceIndex` and gets 1, which tells it to use the second time in the `[20, 55]` repeating list
  * the `quoteAtTime()` return promise resolves to a quote with `time=3n` and a price based on 55
* then the `makeRepeater()` handler fires with `time=3n`, which updates `currentPriceIndex` to 2, and sets `latestTick = 3n`

So the test was asking for the price as of time 3n, but that was a bit ambiguous as to whether it wanted the price before or after the `startTicker` handler had a chance to increase the `currentPriceIndex`. The test behavior depended upon getting the price *before* the increment.

In the new version:
* I changed the `makeRepeater` to start at time=1, not 0, to avoid the extra `wake(time=0)` call from causing the `currentPriceIndex` to be incremented one time too many
  * I also changed the handler to return immediately if `time===0n`, as a backup
* to avoid an ambiguity, I changed `quoteAtTime` to remove the uncoordinated `setWakeup()` call
  * instead, I added a `timeClients` list
  * `quoteAtTime` pushes a `[when, resolver]` pair onto that list
  * the `startTicker` handler updates `currentPriceIndex` and `latestTick`, then walks `timeClients` to find entries where `when >= latestTick`, and calls the resolver
  * this ensures that the `quoteAtTime` action happens consistently *after* the `currentPriceIndex` has been updated

As a result, when the `priceAuthority quoteAtTime` test wants a price of 55, it really needs to ask for the price at time 2n, not 3n. I updated the test to use `quoteAtTime(2n)`, and to expect to see `2n` in the quote's timestamp.

I hit a similar problem in `zoe/test/unitTests/contracts/test-callSpread.js`, which exercises timed options. I had to change the `expiration: 3n` to `2n` so they would sample the expected price.

Most of the other tests are working now that I've completed the mechanical changes of the `buildManualTimer()` signature, and passing in `eventLoopIteration` on the tests that need it.

I'm still trying to figure out a failure in `inter-protocol/test/vaultFactory/test-vaultFactory.js` (in the `mutable liquidity triggers and interest` test). I think it's the same category of failure; it sees a difference in the expected `shortfallBalance`, and I suspect that's because a price is being sampled at the wrong time.

warner commented:
I walked through `test-vaultFactory.js` with @Chris-Hibbert just now. It looks like the old version was doing more `tick()`s than it should have needed, because the old `tick()` was unable to completely wait for the promise queue to drain. So the test had evolved to meet the bugs in the infrastructure. The new `tick()` (in that test, and several others) is configured to use `eventLoopIteration()` to fully drain the queue, and so it needs one fewer tick. I'll put a note in the commit comment with the details.

warner commented:
Ok, everything should be working now. I suggest looking at each commit separately: the first one is just swingset, the second is the less-magical stuff in zoe/governance/inter-protocol that needed updating, and the last two are the more-magical stuff that needed updating. The commit comments contain a deep description of the magic.

The dapp-oracle tests are broken by the changed behaviors. I have a PR to fix it (https://github.com/Agoric/dapp-oracle/pull/40) which I'll land as soon as this one lands.

I still owe changes to https://github.com/Agoric/documentation to update our public docs on this API.

This does not yet enable branded timestamps, nor does it seek to POLA de-power clients from using the full TimerService into using just Clock or TimerBrand. Also the types might be wonky: I updated everything I could figure out, but the main `vat-timer.js` is not yet marked as returning a `{TimerService}` (I tried, but got stuck with something around iterators).

This is marked as merely referring to #4286 (instead of closing it) because `device-timer.js` is left untouched, so the "ever-growing Repeater array" is still there, however now vat-timer doesn't use that feature of the device, so it doesn't matter. I'll re-milestone that ticket once this lands, and a future cleanup pass can strip out everything in the device except the one `setWakeup` that we still need.

Likewise, this refs #5798 instead of closing it because we aren't yet using branded time.

warner commented:
I re-rebased because some zoe durability changes just landed.. wanted to make sure they didn't conflict. Sorry for the churn.

Chris-Hibbert commented:
@warner, this new code should be converted to arrow function style.   WebStorm has good support for converting. Would you like me to push a commit that fixes all the `functions` here?

Chris-Hibbert commented on packages/SwingSet/test/timer/test-timer.js:
```diff
@@ -72,4 +72,48 @@ test('timer vat', async t => {
   t.deepEqual(parse(cd9.body), 'bad setWakeup() handler');
+
+  // start a repeater that should first fire at now+delay+interval, so
+  // 7+20+10=27,37,47,57,..  TODO: poll at 25,35,40
```
Why `TODO`?   The test does poll() 25,35,40.

warner replied:
you're right, that's stale, I'll remove it

Chris-Hibbert commented on packages/SwingSet/test/test-vat-timer.js:
```diff
@@ -0,0 +1,1166 @@
+});
+
+test('clock', async t => {
+  // clock.getCurrentTimestamp() (and nothing else)
```
and `clock.getTimerBrand()`?

Chris-Hibbert commented on packages/inter-protocol/test/vaultFactory/test-vaultFactory.js:
```diff
@@ -782,16 +784,14 @@ test('price falls precipitously', async t => {
 
-  await manualTimer.tick();
+  await manualTimer.tick(); // t 2->3, p 1650->150, liquidates
+  await eventLoopIteration();
```
Isn't this `eventLoopIteration();` unnecessary?

warner replied:
correct, I'll remove it

Chris-Hibbert reviewed (commented):
This looks good.

The main thing it needs is conversion of the `function`s to arrow-style (which I'll happily help with).

I have a couple of question/suggestions; nearly all on comments, though one is on the `eventLoopIteration()` added back in test-vaultFactory.js.

warner commented:
> @warner, this new code should be converted to arrow function style. WebStorm has good support for converting. Would you like me to push a commit that fixes all the `functions` here?

Yeah, I guess so. I find the arrow style less readable, so I haven't been using it in kernel code, but I'd like to see how it looks, and maybe that will get me over it. Let's convert everything in `vat-timer.js`, `manual-timer.js`, `test-vat-timer.js`, and `test-manual-timer.js`.. I think that would avoid single files having a mix of styles, while not increasing the scope of the PR.

warner commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,977 @@
+  }
+
+  /**
+   * delay(delay): return a Promise that fires (with the scheduled
```
Oh, hm, so here the function is defined as `addDelay`, but when exported on `TimerService` it's named `delay`. I changed the local function name to avoid confusion with the `delay` parameter that it takes, and so the reader can correlate it with the external API. Which way do you think I should spell it in the comment? And, would it be a good idea to move this to `const delay = (delay, cancelToken) => ..` instead of `const addDelay = ..` ? The visual confusion would still exist, and I think eslint would still complain about the scope shadowing...

Chris-Hibbert replied:
The comment should reflect the local name.  If you want, it would be fine to add a parenthetical remark that it will be exported as `delay(delay, cancelToken)`.

I don't think `const delay = (delay, ...)` would be better.

warner commented on packages/SwingSet/src/vats/timer/vat-timer.js:
```diff
@@ -1,98 +1,977 @@
+    return { start, interval, active };
+  }
+  const repeaterFacets = {
+    cancel: {}, // internal CancelToken
```
Aye. Do you know of any way to maintain the visual conciseness of:

```js
  const timerService = vivifySingleton(baggage, 'timerService', {
    getCurrentTimestamp,
    setWakeup, // one-shot with handler (absolute)
    wakeAt, // one-shot with Promise (absolute)
    delay: addDelay, // one-shot with Promise (relative)
    makeRepeater, // repeater with Repeater control object (old)
    repeatAfter, // repeater without control object
    makeNotifier, // Notifier
    cancel, // cancel setWakeup/wakeAt/delay/repeat
    getClock: () => timerClock,
    getTimerBrand: () => timerBrand,
  });
```

given that IDE interpretation constraint? I'd find the following less readable, although maybe that's the only way to avoid the problem:
```js
  const timerService = vivifySingleton(baggage, 'timerService', {
    getCurrentTimestamp,
    // one-shot with handler (absolute)
    setWakeup,
    // one-shot with Promise (absolute)
    wakeAt,
    // one-shot with Promise (relative)
    delay: addDelay,
    // repeater with Repeater control object (old)
    makeRepeater,
    // repeater without control object
    repeatAfter,
    // Notifier
    makeNotifier,
    // cancel setWakeup/wakeAt/delay/repeat
    cancel,
    getClock: () => timerClock,
    getTimerBrand: () => timerBrand,
  });
```

Chris-Hibbert replied:
This is as concise. The use of `/* */` comments means that typescript doesn't pick them up, but perhaps that's what you want here.

``` js
  const timerService = vivifySingleton(baggage, 'timerService', {
    getCurrentTimestamp,
    setWakeup /* one-shot with handler (absolute) */,
    wakeAt /* one-shot with Promise (absolute) */,
    delay: addDelay /* one-shot with Promise (relative) */,
    makeRepeater /* repeater with Repeater control object (old) */,
    repeatAfter /* repeater without control object */,
    makeNotifier /* Notifier */,
    cancel /* cancel setWakeup/wakeAt/delay/repeat */,
    getClock: () => timerClock,
    getTimerBrand: () => timerBrand,
  });
```

warner replied:
ah! I hadn't even thought about `/**/` being treated differently. Applied, thanks.

Chris-Hibbert commented:
>> @warner, this new code should be converted to arrow function style. WebStorm has good support for converting. Would you like me to push a commit that fixes all the functions here?

> Yeah, I guess so. ... Let's convert everything in vat-timer.js, manual-timer.js, test-vat-timer.js, and test-manual-timer.js.. I think that would avoid single files having a mix of styles, while not increasing the scope of the PR.

I pushed [b6d8ab6](https://github.com/Agoric/agoric-sdk/pull/5847/commits/b6d8ab60eb3686a8231869817a7c8184e127d591)

> I find the arrow style less readable, so I haven't been using it in kernel code, but I'd like to see how it looks, and maybe that will get me over it. 

I agree about readability, but there were other considerations that we decided mattered even more than readability.

warner commented:
@FUDCo and I walked through this code today, so I think I have a virtual r+ from him. We identified a couple of things to be cleaned up, and I did some refactoring to save a few duplicate `getNow()` device calls. I think I'm ready to land. I'll rebase (to fix that `test-priceAggregator.js` merge conflict) and squash back down into a small number of commits.

FUDCo reviewed (approved):
Go fer it.

turadg commented on packages/zoe/tools/manualTimer.js:
```diff
@@ -1,197 +1,97 @@
+const buildManualTimer = (log = nolog, startValue = 0n, options = {}) => {
+  const {
+    timeStep = 1n,
+    eventLoopIteration = () => 0,
```
when I looked over the timer PR test change I didn’t see anything surprising except that eventLoopIteration has to be passed in as an arg. I remember skimming it and thinking “okay” but I just took another look and I’m confused why it has to be passed in as a function instead of a flag. E.g. `iterateEventLoop: boolean = false`.

@warner what motivates receiving an arbitrary function?

warner replied:
There are a handful of "swingset" unit tests (as opposed to "unit" unit tests) which use the manual timer too, which means they import `manualTimer.js` into code that runs inside a real vat. And `setImmediate` is not available to vat code (the kernel doesn't populate it in the Compartments that houses the vat code) to keep vats from retaining agency after claiming their crank/delivery has finished. So `manualTimer.js` can't provide that functionality by itself.

We experimented with a version that used `if (globalThis.setImmediate)` to try and work with both, but it ended up being unsatisfying, and it wasn't clear that all tests would want (or could even tolerate) the automatic wait-for-flush behavior all the time, so it seemed better to give tests a choice.

That might be worth revisiting, now that we figured out the other test failures (which turned out to be the result of different assumption).

turadg replied:
Ah. I'll humbly ask that we make the common case simpler (without real vat) even if it makes the utility more complex. Consider:
- add default option `iterateEventLoop = false`
- make `eventLoopIteration` option default to undefined
- when `iterateEventLoop` is true, manualTimer does that automatically if it can
- if it can't, it uses `eventLoopIteration` option and throws an error if it's not set, informing the dev that with a real vat it must be passed in
